src/context.rs

```diff
@@ -10,6 +10,306 @@ use crate::ffi::types::{c_uint, c_void, AlignedType};
+    // environment, because if we are pre-empted by an interrupt handler while
+    // the lock is held, and that interrupt handler attempts to take the lock,
+    // then we deadlock.
+    //
+    // Instead, the strategy we take within this module is to simply create a
+    // new stack-local context object if we are unable to obtain a lock on the
+    // global one. This is slow and loses the defense-in-depth "rerandomization"
+    // anti-sidechannel measure, but it is better than deadlocking..
+    struct SpinLock {
+        flag: AtomicBool,
+        // Invariant: if this is non-None, then the store is valid and can be
+        // used with `ffi::secp256k1_context_preallocated_create`.
+        data: UnsafeCell<([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>)>,
+    }
+
+    // Required by rustc if we have a static of this type.
+    // Safety: `data` is accessed only while the `flag` is held.
+    unsafe impl Sync for SpinLock {}
+    unsafe impl Send for SpinLock {}
+
+    impl SpinLock {
+        const fn new() -> Self {
+            Self {
+                flag: AtomicBool::new(false),
+                data: UnsafeCell::new(([AlignedType::ZERO; MAX_PREALLOC_SIZE], None)),
+            }
+        }
+
+        /// Blocks until the lock is acquired, then returns an RAII guard.
+        fn try_lock(&self) -> Option<SpinLockGuard<'_>> {
+            for _ in 0..MAX_SPINLOCK_ATTEMPTS {
+                // `compare_exchange_weak` is fine here: we’re spinning anyway.
+                if self
+                    .flag
+                    .compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+                    .is_ok()
+                {
+                    return Some(SpinLockGuard { lock: self });
+                }
+                spin_loop();
+            }
+            None
+        }
+
+        #[inline(always)]
+        fn unlock(&self) { self.flag.store(false, Ordering::Release); }
+    }
+
+    /// Drops the lock when it goes out of scope.
+    pub struct SpinLockGuard<'a> {
+        lock: &'a SpinLock,
+    }
+
+    impl Deref for SpinLockGuard<'_> {
+        type Target = ([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>);
+        fn deref(&self) -> &Self::Target {
+            // Safe: we hold the lock.
+            unsafe { &*self.lock.data.get() }
+        }
+    }
+
+    impl DerefMut for SpinLockGuard<'_> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            // Safe: mutable access is unique while the guard lives.
+            unsafe { &mut *self.lock.data.get() }
+        }
+    }
+
+    impl Drop for SpinLockGuard<'_> {
+        fn drop(&mut self) { self.lock.unlock(); }
+    }
+
+    /// Borrows the global context and do some operation on it.
+    ///
+    /// If provided, after the operation is complete, [`rerandomize_global_context`]
+    /// is called on the context. If you have some random data available,
```

Comment:
```suggestion
    /// is called on the context. If you have some random data available.
```
In 9330ccde8d57110d75cb6a93db33555fff3aa408 and again in the previous commit for the `std` version.

Reply:
Heh, this sentence is just totally broken. It should say "If some random data is provided, then after the operation is complete, [`rerandomize_global_context`] is called."

Reply:
lol, hopefully one of the other ones only needed the full stop.